[AINFRA-435] Test Setup to Validate Android Agent (`mac-metal`)

ParaskP7 opened the pull request:
Related: #14153

> [!IMPORTANT]
> DO NOT MERGE — This PR is used to test CI scenarios only

This is a draft PR that I will use to (primary) run Android UI tests on the new CI agents (based on `mac-metal`), which is part of [this](https://linear.app/a8c/project/provision-new-bare-metal-ci-agents-981ef3d39744) project.

FYI: Currently this will be used to test [AINFRA-433](https://linear.app/a8c/issue/AINFRA-433) but as the project progress I'm expecting I'll continue to use this dummy PR and branch to test more parts/steps/issues of the project (in the context of Android, mostly UI tests, as well as the generation of Baseline Profiles, see #12429).

AliSoftware committed:
Tweak pipeline to test mac-metal agent with Hello World

ParaskP7 committed:
CI: Add instrumented tests job that runs a specific reviews ui test

ParaskP7 committed:
CI: Temporarily remove setting up gems from the instrumented tests job

As expected this job is currently failing, see below:

:rubygems: Setting up Gems
Using a8c-ci-cache as cache bucket
No cache entry found for 'woocommerce-android-Darwin-arm64-libcmacos-
 15.3.2-ruby3.2.2-xyz'
Fetching bundler-2.5.20.gem
ERROR:  While executing gem ... (Gem::FilePermissionError)
    You don't have write permissions for the /Library/Ruby/Gems/2.6.0
     directory.
🚨 Error: The command exited with status 1
user command error: exit status 1

------------------------------------------------------------------------

Anyway, executing 'gradlew' and the 'connectedDebugAndroidTest'
directly, without setting up any Gems, or installing secrets should be
enough for now and this phase of testing.

ParaskP7 committed:
CI: Add assemble job

AliSoftware commented:
No big deal but FYI I already had https://github.com/woocommerce/woocommerce-android/pull/14153 as a dummy Draft PR to test WCAndroid builds on `mac-metal` 😛 😄

ParaskP7 commented:
> No big deal but FYI I already had https://github.com/woocommerce/woocommerce-android/pull/14153 as a dummy Draft PR to test WCAndroid builds on mac-metal 😛 😄

Of course, this is not news for me @AliSoftware and if you notice from the description itself (see `Related` on top), I based this on your #14153 PR, kept this one `Test mac-metal queue` job you had and then started adding more, Android specific jobs:
1. `Instrumented tests (ReviewsUITest)` first, for [AINFRA-435](https://linear.app/a8c/issue/AINFRA-435/get-a-test-project-ready-to-validate-android-agent).
2. `Assemble` then, for [AINFRA-940](https://linear.app/a8c/issue/AINFRA-940/install-android-sdk).

FYI: I went with this approach in order to not pollute your PR, build logic with any Android specific jobs, and so that we could work on `mac-metal` testing in parallel.

ParaskP7 committed:
CI: Add instrumented tests job with gmd

ParaskP7 committed:
CI: Launch emulator on instrumented tests job without gmd

AliSoftware committed:
Install secrets before running Instrumented Tests

AliSoftware committed:
Provide secrets during Assemble

AliSoftware committed:
Enforce -wipe-data for each CI run

To avoid encountering issues when a previous CI job on the same mac-metal machine already installed a previous APK

AliSoftware committed:
Intentially forgetting to kill the emulator to test if the pre-exit hook we have on mac-metal will properly take care of the safety cleanup

AliSoftware committed:
Add some debugging for emulator/qemu processes still existing at end of script

AliSoftware committed:
More debug info
